getc_putc.8
.TH getc_putc 8
.SH "NAME"
getc_putc \- program to test hard drive performance.

.SH "SYNOPSIS"
.B getc_putc
.I [\-d dir] [\-s size(KiB)] [-m machine-name]
.I [\-u uid\-to\-use:gid\-to\-use] [\-g gid\-to\-use]

.SH "DESCRIPTION"
This manual page documents briefly the
.BR getc_putc ,
program.
.P
This is a simple adjunct to the
.B bonnie++
benchmark.  It is used to test various ways of doing IO one byte at a time,
usually you don't need to do enough of this for it to be a performance issue
for it to matter much which way you do it.  But sometimes it's necessary (for
example whan parsing IO from a terminal and then launching another process
which will take over all IO, such as a simple shell).
.P
The real benefits of this are to help settle some arguements about the
performance of such things, and to educate novices about how bad per-byte IO
really is.

.SH "OPTIONS"
For getc_putc every option is of the form of a hyphen followed by a letter and
then the next parameter contains the value.
.TP
.B \-d
the directory to use for the tests.
.TP
.B \-s
the size of the file for byte IO performance measured in kilobytes.  NB You can
specify the size in mega\-bytes if you add 'm' to the end of the number.

The default for this test is to test with a 512MiB file.  Of the file only 1/64
of it will be used for write() and read() system calls (anything else takes
too long).

.TP
.B \-m
name of the machine \- for display purposes only.

.TP
.B \-u
user\-id to use.  When running as root specify the UID to use for the tests.  It
is not recommended to use root, so if you really want to run as root then use
.B \-u root.
Also if you want to specify the group to run as then use the
.B user:group
format.  If you specify a user by name but no group then the primary group of
that user will be chosen.  If you specify a user by number and no group then
the group will be
.B nogroup.

.TP
.B \-g
group\-id to use.  Same as using
.B :group
for the
.B \-u
parameter, just a different way to specify it for compatibility with other
programs.

.TP
.B \-q
quiet mode. If specified then some of the extra informational messages will be
suppressed. Also the csv data will be the only output on standard out and the
plain text data will be on standard error. This means you can run
.B getc_putc \-q >> file.csv
to record your csv data.

.SH "OUTPUT"
The primary output is plain\-text in 80 columns which is designed to fit well
when pasted into email and which will work well with Braille displays.
.P
The second type of output is CSV (Comma Seperated Values).  This can easily be
imported into any spread\-sheet or database program.
.P
For every test the result is a speed in KiB/s.  I do not display the CPU time
because it presumably is 99% of the power of a single CPU (or something very
close to that).

.SH "AUTHOR"
This program, it's manual page, and the Debian package were written by
Russell Coker <russell@coker.com.au>.
.P
The documentation, the Perl scripts, and all the code for testing the creation
of thousands of files was written by Russell Coker, but the entire package is
under joint copyright with Tim Bray.

.SH "SIGNALS"
Handles SIGINT and does a cleanup (which may take some time), a second SIGINT
or a SIGQUIT will cause it to immidiately die.
.P
SIGXCPU and SIGXFSZ act like SIGINT.
.P
Ignores SIGHUP.

.SH "AVAILABILITY"
The source is available from http://www.coker.com.au/bonnie++ .
.P
See http://etbe.coker.com.au/category/benchmark for further information.

.SH "SEE ALSO"
.BR bonnie++ (8),
.BR zcav (8)


debian/changelog
bonnie++ (2.00a) unstable; urgency=medium

  * Missed the patch in the previous ver, so had to do another.

 -- Russell Coker <russell@coker.com.au>  Tue, 22 Sep 2020 22:37:01 +1000

bonnie++ (2.00) unstable; urgency=medium

  * Fixed an error in assigning colors in bon_csv2html.
  * Called it version 2.00 because it's time.

 -- Russell Coker <russell@coker.com.au>  Tue, 22 Sep 2020 21:35:02 +1000

bonnie++ (1.98) unstable; urgency=medium

  * Fixed macros in bon_csv2html.cpp that had lower case due to excessive
    matching on a regex.
  * Changed debian/compat to level 10
  * Allow specifying the number of random seeks and the number of seeker
    processes and store that in the CSV.
  * Changed bon_csv2html and bon_csv2txt to take the new CSV format.
  * Changed the text output to use KiB/MiB/GiB as units of measurement so we
    can fit NVMe results on screen.
  * Changed the HTML to be in MiB/s for the bulk IO tests.
  * Changed the default size for getc_putc to work with faster CPUs and more
    optimised libc.

 -- Russell Coker <russell@coker.com.au>  Wed, 26 Dec 2018 15:14:28 +1100

bonnie++ (1.97.3) unstable; urgency=medium

  * In bonnie++.8 documented -c option. Closes: #639452 #631484
    Noted that for latency lower is better. Closes: #717285
  * Fixed the symlink creation. Closes: #823882
  * Change backgroundcolor to background-color for correct CSS. Closes: #242878

 -- Russell Coker <russell@coker.com.au>  Tue, 13 Dec 2016 13:37:51 +1100

bonnie++ (1.97.2) unstable; urgency=medium

  * Added compat level and build-depends on debhelper >= 9
    Closes: #817380  Closes: #817313
  * Fix description Closes: #699576
  * Added smoketest Closes: #719739
  * Make it build with GCC-6

 -- Russell Coker <russell@coker.com.au>  Thu, 30 Jun 2016 19:22:23 +1000

bonnie++ (1.97.1) unstable; urgency=medium

  * Fixed the changelog and aim it at unstable not wheezy.

 -- Russell Coker <russell@coker.com.au>  Fri, 23 Nov 2012 20:12:46 +1100

bonnie++ (1.97) unstable; urgency=medium

  * Fix a CSV column alignment bug in the case where -s0 is used.  Bump the
    data format version to 1.97.  Thanks to Brian Keefer for reporting this.
    Also added the concurrency and seed fields to the header.
    Closes: #660564, #669124
  * Use a const pointer for the large file test.
  * zcav.8 documents the maximum value of -s
  * Made the zcav deefault block size 512M to cope with new disks that can
    sustain >200MB/s.  Also documented this.
  * Made zcav not SEGV when the -c option is used.
    Closes: #687668
  * Describe the units for the plain text output in bonnie++.8.
    Closes: #643821
  * Removed the outdated section in the description which referred to Lenny.
    Closes: #693483
  * Fixed most of the HTML validation errors in the output of bon_csv2html.

 -- Russell Coker <russell@coker.com.au>  Fri, 23 Nov 2012 19:32:00 +1100

bonnie++ (1.96) unstable; urgency=low

  * Fix a minor bug in setting the GID.
  * Make it build on Solaris by checking for O_DIRECT being defined.
  * The option -n9765 or greater caused the small file test to abort,
    to increase this I changed the number format to hexadecimal and made it
    10 digits.  Now any option smaller than -n1073741824 will be OK.
    Thanks Nick Dokos of HP.
  * Updated bon_csv2html, made the machine name span two rows, made the
    concurrency column go away when there's no data, correctly calculated the
    colors of the speed columns (was off by one column before), and calculate
    the CPU colors.

 -- Russell Coker <russell@coker.com.au>  Sat, 04 Jul 2009 11:24:13 +1000

bonnie++ (1.95) unstable; urgency=low

  * Removed code for building on OS/2, I think that no-one has been interested
    in this for a while.
  * Removed code for building on NT, I don't think that anyone was ever very
    interested in that and it probably didn't work.
  * Patch from Justin Clift to make getc_putc compile on RHEL 5.2.
  * Added the -D option to bonnie++ to use direct IO (O_DIRECT) for bulk IO.
    Thanks to Dave Murch from Violin Memory - http://violin-memory.com/
  * Fixed some more warnings with recent versions of GCC.

 -- Russell Coker <russell@coker.com.au>  Wed, 31 Dec 2008 08:50:06 +1100

bonnie++ (1.94) unstable; urgency=low

  * Fixed typos in man page.
    Closes: #133714, #449596
  * Added the URL to the man pages.
  * Documented why UID==0 is bad.
    Closes: #182023
  * Check return value of chdir()
  * Removed the zcav -s option which did the same thing as -n.
  * Made zcav report in GiB and MiB/s, and made the default 256MiB.
  * Improved bonnie++.8 documentation of the -y and -p options.
  * Made zcav support skipping blocks for faster completion.
  * Fixed the autoconf checks for 64bit seek.

 -- Russell Coker <russell@coker.com.au>  Tue, 19 Aug 2008 18:36:08 +1000

bonnie++ (1.93d) unstable; urgency=low

  * Documented the default chunk size in bonnie++.8.

  * Added support for testing write speed to zcav.

  * Made default block size for zcav 200M.

 -- Russell Coker <russell@coker.com.au>  Mon,  3 Dec 2007 10:18:00 +1100

bonnie++ (1.93c) unstable; urgency=low

  * Made it compile with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Wed,  8 Jan 2003 23:13:00 +0100

bonnie++ (1.93b) unstable; urgency=low

  * Fixed an error in calculating seeks, it used to add an extra 5 seconds to
    the test time before calculation and therefore under-reported the results.

  * Fixed the signal handling error on Irix.

  * Fixed <algo> check to work with GCC 3.2.

  * Fixed a bug where it would SEGV if you specified more than 999 directories
    (now it handles 99,999 and displays an error message if you ask for more).

  * Fixed a bug in error handling during a stat test with multiple directories.

  * Fixed the mandir for RPM package building.

 -- Russell Coker <russell@coker.com.au>  Wed,  1 Jan 2003 18:02:00 +0100

bonnie++ (1.93a) unstable; urgency=low

  * Made the signal handler extern "C".

  * Updated the RPM packaging.


getc_putc.cpp
#include "bonnie.h"

#include <unistd.h>
#include <sys/utsname.h>
#include <stdlib.h>
#include <cstring>
#include <vector>

#include "duration.h"
#include "getc_putc.h"

static void usage()
{
  fprintf(stderr, "usage:\n"
    "getc_putc [-d scratch-dir] [-s size(KiB)] [-m machine-name]\n"
    "[-u uid-to-use:gid-to-use] [-g gid-to-use]\n"
    "\nVersion: " BON_VERSION "\n");
  exit(eParam);
}

enum getc_tests_t
{
  Write = 0,
  Read,
  PutcNoTh,
  GetcNoTh,
  Putc,
  Getc,
  PutcUnlocked,
  GetcUnlocked,
  GetcTestCount
};

static void print_stat(FILE *fp, double elapsed, int test_size, bool csv);
static void print_all_res(CPCCHAR machine, FILE *fp, double *res, int size, bool csv);

#define WRITE_SIZE_FACT 64

int main(int argc, char *argv[])
{
  int file_size = 512 << 10;
  PCCHAR dir = ".";
  bool quiet = false;
  char *userName = NULL, *groupName = NULL;
  PCCHAR machine = NULL;

  int int_c;
  while(-1 != (int_c = getopt(argc, argv, "d:s:u:g:m:q")) )
  {
    switch(char(int_c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'd':
        dir = optarg;
      break;
      case 's':
        file_size = size_from_str(optarg, "m");
      break;
      case 'q':
        quiet = true;
      break;
      case 'm':
        machine = optarg;
      break;
      case 'g':
        if(groupName)
          usage();
        groupName = optarg;
      break;
      case 'u':
      {
        if(userName)
          usage();
        userName = strdup(optarg);
        int i;
        for(i = 0; userName[i] && userName[i] != ':'; i++) {}

        if(userName[i] == ':')
        {
          if(groupName)
            usage();
          userName[i] = '\0';
          groupName = &userName[i + 1];
        }
      }
      break;
    }
  }

  if(userName || groupName)
  {
    if(bon_setugid(userName, groupName, quiet))
      return 1;
    if(userName)
      free(userName);
  }
  else if(geteuid() == 0)
  {
    fprintf(stderr, "You must use the \"-u\" switch when running as root.\n");
    usage();
  }

  if(machine == NULL)
  {
    struct utsname utsBuf;
    if(uname(&utsBuf) != -1)
      machine = utsBuf.nodename;
  }

  file_size -= (file_size % WRITE_SIZE_FACT);
  file_size = file_size << 10;
  if(!file_size)
    usage();

  char *fname = new char[22 + strlen(dir)];

  sprintf(fname, "%s/getc_putc.%d", dir, getpid());

  int fd = open(fname, O_CREAT | O_TRUNC | O_RDWR, S_IRUSR | S_IWUSR);
  if(fd < 0)
  {
    fprintf(stderr, "Can't create file \"%s\".\n", fname);
    usage();
  }
  if(dup2(fd, FILE_FD) != FILE_FD)
  {
    fprintf(stderr, "Can't dup2() the file handle.");
    return 1;
  }
  close(fd);

  if(!quiet)
    printf("Extending file...");
  fflush(NULL);
  char buf[1 << 20];

  int size = 0, wrote;
  while(size < file_size)
  {
    wrote = write(FILE_FD, buf, min(sizeof(buf), (size_t)file_size - size));
    if(wrote < 0)
    {
      fprintf(stderr, "Can't extend file - disk full?\n");
      return 1;
    }
    size += wrote;
  }
  fsync(FILE_FD);
  volatile char c;
  int i;
  Duration dur;
  double res[GetcTestCount];

  if(lseek(FILE_FD, 0, SEEK_SET) != 0)
  {
    fprintf(stderr, "Can't seek.\n");
    return 1;
  }
  size = file_size / WRITE_SIZE_FACT;
  TEST_FUNC_WRITE("write(fd, &c, 1)", if(write(FILE_FD, (void *)&c, 1) != 1), res[Write]);
  fsync(FILE_FD);
  if(lseek(FILE_FD, 0, SEEK_SET) != 0)
  {
    fprintf(stderr, "Can't seek.\n");
    return 1;
  }
  TEST_FUNC_READ("read(fd, &c, 1)", if(read(FILE_FD, (void *)&c, 1) != 1), res[Read]);

  char *prog = new char[strlen(argv[0]) + 30];
  sprintf(prog, "%s_helper %d", argv[0], file_size);
  if(quiet)
    strcat(prog, "q");
  FILE *child = popen(prog, "r");
  if(!child)
  {
    fprintf(stderr, "Can't execute \"%s\".\n", prog);
    return 1;
  }
  if(fread(&res[PutcNoTh], sizeof(double) * 2, 1, child) != 1)
  {
    fprintf(stderr, "Can't get results from child.\n");
    return 1;
  }
  fclose(child);

  FILE *fp = fdopen(FILE_FD, "w+");
  if(!fp)
  {
    fprintf(stderr, "Can't reopen for putc.\n");
    return 1;
  }
  if(fseek(fp, 0, SEEK_SET) != 0)
  {
    fprintf(stderr, "Can't seek.\n");
    return 1;
  }
  fflush(NULL);
  size = file_size;
  TEST_FUNC_WRITE("putc(c, fp)", if(putc(c, fp) == EOF), res[Putc]);
  if(fseek(fp, 0, SEEK_SET) != 0)
  {
    fprintf(stderr, "Can't seek.\n");
    return 1;
  }
  fflush(NULL);
  TEST_FUNC_READ("getc()", if( (c = getc(fp)) == EOF), res[Getc]);
  if(fseek(fp, 0, SEEK_SET) != 0)
  {
    fprintf(stderr, "Can't seek.\n");
    return 1;
  }
  fflush(NULL);
  size = file_size;
  TEST_FUNC_WRITE("putc_unlocked(c, fp)", if(putc_unlocked(c, fp) == EOF), res[PutcUnlocked]);
  if(fseek(fp, 0, SEEK_SET) != 0)
  {
    fprintf(stderr, "Can't seek.\n");
    return 1;
  }
  fflush(NULL);
  TEST_FUNC_READ("getc_unlocked()", if( (c = getc_unlocked(fp)) == EOF), res[GetcUnlocked]);

  if(!quiet)
    printf("done\n");
  fclose(fp);
  unlink(fname);
  size = size / 1024;
  print_all_res(machine, stderr, res, size, false);
  print_all_res(machine, stdout, res, size, true);

  return 0;
}

static void print_all_res(CPCCHAR machine, FILE *fp, double *res, int size, bool csv)
{
  if(!csv)
  {
    fprintf(fp, "Version %5s          write   read putcNT getcNT   putc   getc  putcU  getcU\n", BON_VERSION);
    fprintf(fp, "%-20s ", machine);
  }
  else
  {
    fprintf(fp, "%s", machine);
  }
  print_stat(fp, res[Write], size / WRITE_SIZE_FACT, csv);
  print_stat(fp, res[Read], size / WRITE_SIZE_FACT, csv);
  print_stat(fp, res[PutcNoTh], size, csv);
  print_stat(fp, res[GetcNoTh], size, csv);
  print_stat(fp, res[Putc], size, csv);
  print_stat(fp, res[Getc], size, csv);
  print_stat(fp, res[PutcUnlocked], size, csv);
  print_stat(fp, res[GetcUnlocked], size, csv);
  fprintf(fp, "\n");
}

static void print_stat(FILE *fp, double elapsed, int test_size, bool csv)
{
  if(elapsed == 0.0)
  {
    if(!csv)
      fprintf(fp, "       ");
    else
      fprintf(fp, ",");
  }
  else if(elapsed < MinTime)
  {
    if(!csv)
      fprintf(fp, " ++++++");
    else
      fprintf(fp, ",++++++");
  }
  else
  {
    double speed = double(test_size) / elapsed;
    if(!csv)
      fprintf(fp, " %6d", int(speed));
    else
      fprintf(fp, ",%d", int(speed));
  }
}
